packages/modules/vehicles/smarteq/socutils.py

```diff
@@ -0,0 +1,221 @@
+            client.on_connect = mqtt_on_connect
+            client.on_message = mqtt_on_message
+
+            client.connect("localhost", 1883, 60)
+
+            client.loop_start()
+            time.sleep(2)
+            client.disconnect()
+            client.loop_stop()
+        except Exception as e:
+            log.exception("_get_vehicleList mqtt error: %s", e)
+
+        _infoLog('Q', '_get_vehicleList: vehicleList=\n' + json.dumps(vehicleList, indent=4))
+        _infoLog('Q', '_get_vehicleList: vehicleConfig=\n' + json.dumps(vehicleConfig, indent=4))
+
+        return vehicleList, vehicleConfig
+
+    def _get_logFilter(self) -> str:
+        global logFilter
+        try:
+            logFilter = ''
+            msg = subscribe.simple("openWB/vehicle/3/soc_module/config", hostname="localhost")
+            config = json.loads(str(msg.payload, 'utf-8'))
+            if 'logFilter' in config['configuration']:
+                logFilter = config['configuration']['logFilter']
+                if logFilter is None:
+                    logFilter = ''
+        except Exception as e:
+            log.exception("_get_logFilter mqtt error: %s", e)
+        _infoLog('Q', '_get_logFilter: logFilter=' + logFilter)
+        return logFilter
+
+    # publish a single message to 1883 broker
+    def pub(self, topic: str, payload: str):
+        try:
+            _infoLog('q', "pub: topic=" + topic + ', payload=' + payload)
+            client = mqtt.Client()
+            client.connect("localhost", 1883, 60)
+            ret, mid = client.publish(topic, payload)
+            if ret != 0:
+                _infoLog('q', "pub: topic=" + topic +
+                         ', payload=' + payload +
+                         ', ret=' + str(ret) +
+                         ', mid=' + str(mid))
+            client.disconnect()
+        except Exception as e:
+            log.exception("pub mqtt error: %s", e)
+
+    # publish CarState via MQTT to 1883 broker
+    def set_CarState(self, ev: int, soc: int, range: float, ts: int):
+        try:
+            msgs = []
+            topic = 'openWB/set/vehicle/' + str(ev) + '/get/soc'
+            # self.pub(topic, str(soc))
+            msgs.append({'topic': topic, 'payload': str(soc)})
+            topic = 'openWB/set/vehicle/' + str(ev) + '/get/range'
+            # self.pub(topic, str(range))
+            msgs.append({'topic': topic, 'payload': str(range)})
+            topic = 'openWB/set/vehicle/' + str(ev) + '/get/soc_timestamp'
+            _ts = time.time()
+            # self.pub(topic, str(_ts))
+            msgs.append({'topic': topic, 'payload': str(_ts)})
+            _infoLog('Q', "set_CarState publish.multiple: msgs=\n" + json.dumps(msgs))
+            _infoLog('q', "set_CarState: " + str(soc) + '%/' + str(range) + 'KM@' + str(_ts))
+            publish.multiple(msgs, hostname='localhost', port=1883)
+        except Exception as e:
+            log.exception("set_CarState error: %s", e)
```

Comment:
Bitte nur den CarState zurückgeben und Publishen, Fehlerbehandlung etc vom Backend erledigen lassen. Wenn jedes Modul seine eigene Implementierung nutzt, wird die Wartung des Codes schnell sehr sehr aufwändig.

Reply:
Das Return-Interface püass für das im smarteq benutzte Mercedes-API halt nicht.
Das SOC-Modul subscribed die Daten über Websockets.
Die Fahrzeug-Daten kommen asynchron über WebSockets ca. alle 15 Sekunden. 
Wenn sich Soc oder range ändern schreibt das Modul die Daten über diese Funktion in den CarState.

Ich habe mittlerweile eine Alternative über den store:
    # publish CarState via core store
    def set_CarState(self, vehicle: int, soc: int, range: float, ts: int):
        try:
            st: CarValueStoreBroker = store.get_car_value_store(vehicle)
            carState: CarState = CarState(int(soc), float(range), round(time(), 2))
            st.set(carState)
            st.update()
        except Exception as e:
            log.exception("set_CarState Error: %s", e)
        return

Wäre das ein akzeptabler Weg?

Reply:
Ich habe jetzt die "API" über den store implementiert wie oben.